### ISSUE TITLE
Ability to record "SSL Handshake Failures" as counter in stick tables for src

### ISSUE THREAD

**Issue by amsnek** [haproxy/haproxy#2717]
### Your Feature Request

It would be usefull to have the ability to record the amount of SSL Handshake Failures a src ip produces in a stick table. Currently this can only be extracted from logging mechanisms.

### What are you trying to do?

This would offer the possibility to create ACLs and mitigations based on for example high numbers of SSL Handshake Failures/Attacks. This would help (automating) tcp-connection drops with attackers that utilize a huge range of rotating IP-Adresses. Extracting from logs is delayed and difficult to automate. 

### Output of `haproxy -vv`

```plain
HAProxy version 3.0.4-7a59afa 2024/09/03 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2029.
Known bugs: http://www.haproxy.org/bugs/bugs-3.0.4.html
Running on: Linux 4.18.0-553.16.1.el8_10.x86_64 #1 SMP Thu Aug 1 04:16:12 EDT 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CC      = cc
  CFLAGS  = -O2 -g -fwrapv -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Werror=format-security
  OPTIONS = USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1 USE_NS=1 USE_SYSTEMD=1 USE_QUIC=1 USE_PROMEX=1 USE_STATIC_PCRE2=1 USE_PCRE2_JIT=1 USE_QUIC_OPENSSL_COMPAT=1
  DEBUG   =

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC +QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE +STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=2).
Built with OpenSSL version : OpenSSL 3.3.2 3 Sep 2024
Running on OpenSSL version : OpenSSL 3.3.2 3 Sep 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.7
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.44 2024-06-07
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 8.5.0 20210514 (Red Hat 8.5.0-22)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by wtarreau** [comment]
I understand the idea, I'm just embarrassed that it would be extremely specific to the frontend, while we try to limit our addition of new data types in stick tables to very generic stuff.

However, since we're already storing the "glitch_cnt" counter and "glitch_rate" frequency counter, and these are specifically updated for anything not illegal but suspicious, maybe we should increment them on frontend handshake errors. We're already doing that for invalid H2 prefaces for example. This way you would have the counter you're looking for.

At the moment the glitch counter is stored in the mux (this allows to count separately for frontend and backend), but we already have a function to increment the tracked counters for the session, so we'd just have to do this in the ssl_sock code at the various places we're setting CO_ER_SSL_* for errors, as well as in the heartbeat detection:
```
   if (!conn_is_back(conn))
       session_add_glitch_ctr(conn->owner, 1);
```
and that would automatically do the job. We could even increment the value faster for certain errors. The principle of the glitch counter is that normally you should not worry about small values because they may happen from time to time even with some poorly implemented clients. But when the number increases fast (particularly the glitch_rate) then there's no doubt anymore and you can use a high threshold to decide that some source is really misbehaving.

Is this something you would be interesting in trying by yourself to see if that's sufficient for you ? It should literally be no more than the above at a few selected places I think.

**Comment by amsnek** [comment]
Thank you for putting thought into this! Using "glitch_cnt" or "glitch_rate" for this would certainly be sufficient for my intended purpose. Maybe it could make sense to have an option to enable/disable the counting of things like ssl_handshake_errors count though I dont see a reason why these shouldnt be included. 

An option to change the increment rate (slower/faster) could be interesting, this would enable to put more or less weight on for example the ssl_handshake_error. I certainly dont want to make things more complicated implementation wise as needed though. 

I would definitely be willing to try this myself and test this out! I think this could be a valuable tool, albeit admittedly in somewhat more rare usecases. I recently encountered such a (rare) scenario, which made me realise that I have no real tool against it.

**Comment by wtarreau** [comment]
Rest assured you're not alone with these needs. Regarding the ability to score these events, that's what I initially came up with for H2, but I ended up using 1 for every event type for now to keep things simple. The rationale here is that you really don't care about the exact values, when you're under attack, your scores go x100 or x1000 anyway. The only real benefit in having some configurable scores is to permit 0 in order to ignore certain events. For H2 I even wanted to support negative scores for certain events indicating a legitimate use, but that's getting even trickier. In the case above we could start with 1 (or 10) by default for each SSL handshake error, and see if some of them are too problematic and if a few really need to be made configurable (or maybe we'd configure them as a whole).

If you give it a try, please really share your observations here, as they will help design something good enough for most use cases.

**Comment by amsnek** [comment]
I agree that a multiplier of 1 would be sufficient and keep things much simpler. There may be some edge cases were one could want to weigh one thing more then others but those would be solvable with the mulitplier of 1 as well. 

I agree that exact values are not too relevant since the trend is a much more usefull aspect
. 
I recently observed a larger ddos, that used a lot of IP-Adresses in a rotating manner, many of them just producing handshake failures. Most IP-Adresses only produced a low-ish amount of failures (a dozen up to a few hundred) in relative short order before being rotated, others tens of thousands. But even if their distrubtion and rotation increased and they were to use less failures per ip/time, in such a case, one could (temporary) apply much stricter glitch rules for the affected destinations to eliminate even those attackers and interrupt them as early as possible. Adding them to an abuse list that sticks for a longer period (~24 hours or similiar) could prevent them entirely on their next rotation as well.

**Comment by wtarreau** [comment]
Yes that's totally right. For such cases of handshake failures, you can play with the silent-drop rules as soon as the connection is set up. It will maintain some resources in use on the client and might possibly even block it depending on how it's written (e.g. script in a "for" loop). Make sure to read the doc, though, to take care of your upstream equipments just in case there are any!

**Comment by amsnek** [comment]
Yes, I am currently using silent-drop, initially reject, after a certain amount of observed evilness, that is very neat! Having the ability to set the rst-ttl hop count is important indeed.
